feat(ingestion): add ingestion warning if event more than 23 hours in the future

xvello opened the pull request:
## Problem

We accept events in the future. Although they don't show up in the live events table, we had users confused by persons created in the future by these events.

We decided to not drop these events, but we're adding an ingestion warning if timestamps are more than 23h in the future, to help diagnose instrumentation issues.

![image](https://user-images.githubusercontent.com/18740659/210829203-d57474f3-9537-4583-af01-bd2115407488.png)

## Changes

- Update `parseEventTimestamp` and its ingestion warning callback
- Create new `event_timestamp_in_future` ingestion warning and add it to the UI

## How did you test this code?

Unit testing + local run

xvello committed:
feat(ingestion): add ingestion warning if event more than 23 hours in the future

tomasfarias committed:
test: Update unit tests for ingestion timestamp

hazzadous commented on frontend/src/scenes/data-management/ingestion-warnings/IngestionWarningsView.tsx:
```diff
@@ -73,6 +74,26 @@ const WARNING_TYPE_RENDERER = {
+        }
+        return (
+            <>
+                Timestamp computed to <code>{details.result}</code> from the following input:
```
Is there anywhere we can highlight how and why result is calculated the way that it is? Just a comment around time skew etc. I found it very confusing when I was looking at this first.

xvello replied:
Definitely. We have an autogenerated `docs` links to `https://posthog.com/manual/data-management#an-event-was-sent-more-than-23-hours-in-the-future`, we need to contribute that documentation change. That's where we'll explain this.

hazzadous replied:
👍

hazzadous commented on frontend/src/scenes/data-management/ingestion-warnings/IngestionWarningsView.tsx:
```diff
@@ -73,6 +74,26 @@ const WARNING_TYPE_RENDERER = {
+            <>
+                Timestamp computed to <code>{details.result}</code> from the following input:
+                <ul>
+                    {details.timestamp ? <li>timestamp: {details.timestamp}</li> : ''}
```
```suggestion
                    {details.timestamp ? <li>Client provided timestamp: {details.timestamp}</li> : ''}
```

hazzadous commented on frontend/src/scenes/data-management/ingestion-warnings/IngestionWarningsView.tsx:
```diff
@@ -73,6 +74,26 @@ const WARNING_TYPE_RENDERER = {
+                Timestamp computed to <code>{details.result}</code> from the following input:
+                <ul>
+                    {details.timestamp ? <li>timestamp: {details.timestamp}</li> : ''}
+                    {details.sentAt ? <li> sent_at: {details.sentAt}</li> : ''}
```
```suggestion
                    {details.sentAt ? <li>Client provided sent_at: {details.sentAt}</li> : ''}
```

hazzadous commented on frontend/src/scenes/data-management/ingestion-warnings/IngestionWarningsView.tsx:
```diff
@@ -73,6 +74,26 @@ const WARNING_TYPE_RENDERER = {
+                    {details.timestamp ? <li>timestamp: {details.timestamp}</li> : ''}
+                    {details.sentAt ? <li> sent_at: {details.sentAt}</li> : ''}
+                    {details.offset ? <li>offset: {details.offset}</li> : ''}
+                    <li>capture time: {details.now}</li>
```
```suggestion
                    <li>PostHog server capture time: {details.now}</li>
```

xvello committed:
Apply suggestions from code review

Co-authored-by: Harry Waye <harry@posthog.com>

xvello commented on frontend/src/scenes/data-management/ingestion-warnings/IngestionWarningsView.tsx:
```diff
@@ -73,6 +74,26 @@ const WARNING_TYPE_RENDERER = {
+                <ul>
+                    {details.timestamp ? <li>Client provided timestamp: {details.timestamp}</li> : ''}
+                    {details.sentAt ? <li>Client provided sent_at: {details.sentAt}</li> : ''}
+                    {details.offset ? <li>offset: {details.offset}</li> : ''}
```
```suggestion
                    {details.offset ? <li>Client provided time offset: {details.offset}</li> : ''}
```

xvello committed:
Update frontend/src/scenes/data-management/ingestion-warnings/IngestionWarningsView.tsx

Commit:
Merge remote-tracking branch 'origin/master' into futuretimestamp-iwarning

tomasfarias committed:
test: Use a timestamp only 23h in the future

This allows us to test the exact boundary of the
'event_timestamp_in_future' warning to ensure it doesn't pop.
